homeassistant/components/binary_sensor/isy994.py

```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
 
 
 class ISYBinarySensorDevice(isy.ISYDevice, BinarySensorDevice):
-    """Representation of an ISY994 binary sensor device."""
+    """Representation of an ISY994 binary sensor device.
+
+    Often times, a single device is represented by multiple nodes in the ISY,
+    allowing for different nuances in how those devices report their on and
+    off events. This class turns those multiple nodes in to a single Hass
+    entity and handles both ways that ISY binary sensors can work.
+    """
 
     def __init__(self, node) -> None:
         """Initialize the ISY994 binary sensor device."""
         isy.ISYDevice.__init__(self, node)
+        self._child_nodes = []
+        self._off_node = None
+        self._heartbeat_node = None
+        self._heartbeat_timestamp = None
+        self._device_class_from_type = self._detect_device_type()
+        # pylint: disable=protected-access
+        self._computed_state = bool(self._node.status._val)
+        node.controlEvents.subscribe(self._positive_node_control_handler)
+
+    def _detect_device_type(self) -> str:
+        try:
+            device_type = self._node.type
+        except AttributeError:
+            self._device_class_from_type = None
+            return None
+
+        split_type = device_type.split('.')
+        _LOGGER.debug(split_type)
+        for cls, ids in ISY_DEVICE_TYPES.items():
+            _LOGGER.debug(split_type[0] + '.' + split_type[1])
+            if split_type[0] + '.' + split_type[1] in ids:
+                self._device_class_from_type = cls
+                return cls
+
+        self._device_class_from_type = None
+        return None
+
+    def add_child_node(self, child):
+        """Add a child node to this binary sensor device.
+
+        The child node can either be a node that receives the 'off' events, or
+        a heartbeat node for reporting that this device is still alive
+        """
+        subnode_id = int(child.nid[-1])
+        if subnode_id == 2:
+            # "Negative" node that can be used to represent a negative state
+            # when it reports "On"
+            child.controlEvents.subscribe(self._negative_node_control_handler)
+            self._off_node = child
+        elif subnode_id == 4:
+            # Heartbeat node that just reports "On" every 24 hours
+            child.controlEvents.subscribe(self._heartbeat_node_control_handler)
+            self._heartbeat_timestamp = STATE_UNKNOWN
```

Comment:
It's important for there to be a difference between the attribute not existing (None) and the attribute currently being "Unknown" because we haven't seen a heartbeat since the last boot. If it was simply None until the first heartbeat, an automation can't tell if this device ever SHOULD give a heartbeat.